presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToJsonCast.java

```diff
@@ -71,45 +68,40 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         Type valueType = boundVariables.getTypeVariable("V");
+        Type mapType = new MapType(keyType, valueType);
+        checkCondition(canCastToJson(mapType), INVALID_CAST_ARGUMENT, "Cannot cast %s to JSON", mapType);
 
-        MethodHandle methodHandle = METHOD_HANDLE.bindTo(keyType);
-        methodHandle = methodHandle.bindTo(valueType);
+        JsonUtil.MapObjectKeyReader reader = JsonUtil.MapObjectKeyReader.createMapObjectKeyReader(keyType);
+        JsonUtil.JsonGeneratorWriter writer = JsonUtil.JsonGeneratorWriter.createJsonGeneratorWriter(valueType);
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(reader).bindTo(writer);
 
         return new ScalarFunctionImplementation(false, ImmutableList.of(false), methodHandle, isDeterministic());
     }
 
     @UsedByGeneratedCode
-    public static Slice toJson(Type keyType, Type valueType, ConnectorSession session, Block block)
+    public static Slice toJson(JsonUtil.MapObjectKeyReader reader, JsonUtil.JsonGeneratorWriter writer, ConnectorSession session, Block block)
     {
-        Map<Object, Object> map = new HashMap<>();
-        for (int i = 0; i < block.getPositionCount(); i += 2) {
-            map.put(JsonFunctions.getJsonObjectValue(keyType, session, block, i), JsonFunctions.getJsonObjectValue(valueType, session, block, i + 1));
-        }
         try {
-            return Slices.utf8Slice(OBJECT_MAPPER.get().writeValueAsString(map));
-        }
-        catch (JsonProcessingException e) {
-            throw Throwables.propagate(e);
-        }
-    }
-
-    // Unfortunately this has to be a raw Map, since Map<?, ?> doesn't seem to work in Jackson
-    private static class MapSerializer
-            extends JsonSerializer<Map>
-    {
-        @Override
-        public void serialize(Map map, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
-                throws IOException
-        {
-            Map<String, Object> orderedMap = new TreeMap<>();
-            for (Map.Entry<?, ?> entry : ((Map<?, ?>) map).entrySet()) {
-                orderedMap.put(entry.getKey().toString(), entry.getValue());
+            Map<String, Integer> orderedKeyToValuePosition = new TreeMap<>();
+            for (int i = 0; i < block.getPositionCount(); i += 2) {
+                String objectKey = reader.read(block, i);
+                orderedKeyToValuePosition.put(objectKey, i + 1);
             }
-            jsonGenerator.writeStartObject();
-            for (Map.Entry<String, Object> entry : orderedMap.entrySet()) {
-                jsonGenerator.writeObjectField(entry.getKey(), entry.getValue());
+
+            SliceOutput output = new DynamicSliceOutput(100);
+            try (JsonGenerator jsonGenerator = createJsonGenerator(JSON_FACTORY, output)) {
+                jsonGenerator.writeStartObject();
+                for (Map.Entry<String, Integer> entry : orderedKeyToValuePosition.entrySet()) {
+                    int valuePosition = entry.getValue();
```

Comment:
@haozhun : Is the convention to avoid create variable as much as possible?

Here I kind of use the variable name as  an comment :).